piker/brokers/ib.py

```diff
@@ -1183,112 +1201,217 @@ def pack_position(pos: Position) -> Dict[str, Any]:
+                # https://www.python.org/dev/peps/pep-0526/#global-and-local-variable-annotations
+                trade: Trade
+                fill: Fill
+                trade, fill = item
+                execu: Execution = fill.execution
+
+                # TODO: normalize out commissions details?
+                details = {
+                    'contract': asdict(fill.contract),
+                    'execution': asdict(fill.execution),
+                    'commissions': asdict(fill.commissionReport),
+                }
 
                 # supposedly IB server fill time
-                'broker_time': execu.time,  # converted to float by us
-                # ns from main TCP handler by us inside ``ib_insync`` override
-                'time': fill.time,
-                'time_ns': time.time_ns(),  # cuz why not
-                'action': action_map[execu.side],
-                'size': execu.shares,
-                'price': execu.price,
-            }
-
-        elif event_name == 'error':
-            msg = item
-
-            # f$#$% gawd dammit insync..
-            con = msg['contract']
-            if isinstance(con, Contract):
-                msg['contract'] = asdict(con)
-
-            if msg['reqid'] == -1:
-                log.error(pformat(msg))
-
-            # don't forward, it's pointless..
-            continue
+                details['broker_time'] = execu.time
+                details['name'] = 'ib'
 
-        elif event_name == 'position':
-            msg = pack_position(item)
+                msg = BrokerdFill(
+                    # should match the value returned from `.submit_limit()`
+                    reqid=execu.orderId,
+                    time_ns=time.time_ns(),  # cuz why not
 
-        if msg.get('reqid', 0) < -1:
-            # it's a trade event generated by TWS usage.
-            log.warning(f"TWS triggered trade:\n{pformat(msg)}")
+                    action=action_map[execu.side],
+                    size=execu.shares,
+                    price=execu.price,
 
-            msg['reqid'] = 'tws-' + str(-1 * msg['reqid'])
+                    broker_details=details,
+                    # XXX: required by order mode currently
+                    broker_time=details['execution']['time'],
 
-            # mark msg as from "external system"
-            # TODO: probably something better then this..
-            msg['external'] = True
+                )
 
-            yield {'remote_trades': (event_name, msg)}
-            continue
+            elif event_name == 'error':
+
+                err: dict = item
+
+                # f$#$% gawd dammit insync..
+                con = err['contract']
+                if isinstance(con, Contract):
+                    err['contract'] = asdict(con)
+
+                if err['reqid'] == -1:
+                    log.error(f'TWS external order error:\n{pformat(err)}')
+
+                # don't forward for now, it's unecessary.. but if we wanted to,
```

Comment:
we do need to relay this in some cases where ib isn't sending a "cancelled" status when there's an order submission failure... it's like weird cases though that i haven't yet categorized.